### PR TITLE
feat(discovery): add consul support

### PR DESCRIPTION
**Pull request by water1st**
### What this PR does / why we need it:

add consul support

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

**Review by spacewander — COMMENTED** [review]
Could you make the current test pass, and add test for the consul discovery like https://github.com/apache/apisix/blob/master/t/discovery/nacos2.t?

You can learn how to write tests with:
https://github.com/apache/apisix/blob/master/docs/en/latest/internal/testing-framework.md
https://metacpan.org/pod/Test%3A%3ANginx%3A%3ASocket

We can start a consul instance in the CI like https://github.com/apache/apisix/blob/82d17abf9987528856cbd20ce5f17f2ae0a799ce/ci/pod/docker-compose.yml#L119

**Comment by spacewander** [comment]
@bisakhmondal 
Could you help @water1st if she/he has trouble with the test cases? Thanks!

File: apisix/discovery/consul/init.lua [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    end
+
+    if consul_conf.alc_token ~= nil and #consul_conf.alc_token ~= "" then
+        path = path .. split .. "token=" .. consul_conf.alc_token
+        if split == "?" then
+            split = "&"
+        end
+    end
+
+    local endpoint = server .. path
```
Comment (bisakhmondal) [inline apisix/discovery/consul/init.lua:54]: You don't have to do it manually. How about using https://github.com/golgote/neturl#querystring-builder ?

**Review by bisakhmondal — COMMENTED** [review]
Hey, thanks for the PR

**Comment by bisakhmondal** [comment]
> @bisakhmondal Could you help @water1st if she/he has trouble with the test cases? Thanks!

Hi @water1st, for the code lint: you can run `make lint` before pushing your changes to the PR ([ref](https://github.com/apache/apisix/blob/master/CONTRIBUTING.md#check-code-style-and-test-case-style))

As you have created a new directory (/discovey/consul), it won't be installed automatically unless you add an entry here. could you please do it? The other CIs will also pass after that.
https://github.com/apache/apisix/blob/bbebdc9606e47f325180737cabf3cf479b69d013/Makefile#L275-L282

**Comment by ZUOXIANGE** [comment]
Why this pr closed?

**Comment by bisakhmondal** [comment]
Hi @water1st, any progress on this PR after the review?
Do let us know if you are stuck anywhere. We would be very happy to help you. Thank you!

**Pull request closed without merging**